### ISSUE TITLE
HTC Desire 510 deodex generates invalid call

### ISSUE THREAD

**Issue by HomerSp** [JesusFreke/smali#472]
I've been trying to deodex an HTC Desire 510 system using both 2.2b4 and 2.1.3, and while the deodex process finished successfully I am not able to boot with the new files. The problem seems to be in core.jar, java/lang/reflect/Field.smali. Specifically, the equals method:
```
.method public equals(Ljava/lang/Object;)Z
    .registers 4
    .param p1, "object"    # Ljava/lang/Object;

    .prologue
    .line 236
    instance-of v0, p1, Ljava/lang/reflect/Field;

    if-eqz v0, :cond_14

    #disallowed odex opcode
    #invoke-virtual-quick {p0}, vtable@7
    nop

    move-result-object v0

    #disallowed odex opcode
    #invoke-virtual-quick {p1}, vtable@7
    nop

    move-result-object v1

    #disallowed odex opcode
    #execute-inline {v0, v1}, inline@3
    nop

    move-result v0

    if-eqz v0, :cond_14

    const/4 v0, 0x1

    :goto_13
    return v0

    :cond_14
    const/4 v0, 0x0

    goto :goto_13
.end method
```

The #invoke-virtual-quick {p1}, vtable@7 line seems to be resolved to
invoke-virtual {p1}, Ljava/lang/reflect/Field;->toString()Ljava/lang/String;
while it should be:
invoke-virtual {p1}, Ljava/lang/Object;->toString()Ljava/lang/String;
causing the optimisation process to fail.

Here's the relevant section from the log:
```
W/dalvikvm(  353): VFY: 'this' arg 'Ljava/lang/Object;' not instance of 'Ljava/lang/reflect/Field;'
W/dalvikvm(  353): VFY:  rejecting opcode 0x6e at 0x0008
W/dalvikvm(  353): VFY:  rejected Ljava/lang/reflect/Field;.equals (Ljava/lang/Object;)Z
W/dalvikvm(  353): Verifier rejected class Ljava/lang/reflect/Field;
E/dalvikvm(  353): ERROR: tried to execute code in unprepared class 'Ljava/lang/reflect/Field;' (-1)
I/dalvikvm(  353): "main" prio=5 tid=1 RUNNABLE
I/dalvikvm(  353):   | group="main" sCount=0 dsCount=0 obj=0x41994f30 self=0x419839f0
I/dalvikvm(  353):   | sysTid=353 nice=0 sched=0/0 cgrp=default handle=1074889112
I/dalvikvm(  353):   | state=R schedstat=( 0 0 0 ) utm=32 stm=18 core=2
I/dalvikvm(  353):   at java.lang.reflect.Field.<init>(Field.java:-1)
I/dalvikvm(  353):   at java.lang.Class.getDeclaredField(Native Method)
I/dalvikvm(  353):   at java.lang.Class.getDeclaredField(Class.java:594)
I/dalvikvm(  353):   at java.util.concurrent.atomic.AtomicInteger.<clinit>(AtomicInteger.java:32)
I/dalvikvm(  353):   at android.os.StrictMode.<clinit>(StrictMode.java:-1)
I/dalvikvm(  353):   at dalvik.system.NativeStart.main(Native Method)
I/dalvikvm(  353):   at dalvik.system.NativeStart.main(Native Method)
```

If needed I can upload the framework files.

**Comment by HomerSp** [comment]
Just tried baksmali 2.0.5, and it seems to be working as expected there.

Should probably mention that this is Android 4.4.2, API level 19.

**Comment by JesusFreke** [comment]
`#disallowed odex opcode` comments are generated when you're disassembling (not deodexing) a dex/odex file with optimized opcodes.

Please provide the full command line you're using.

**Comment by HomerSp** [comment]
The code above is indeed the disassembled odex file. I wanted to show what the original, unoptimized code looked like.

Here's java/lang/reflect/Field.smali when disassembling the odex (using d):
https://gist.github.com/29e20703ba1f11ccb8d7827608b6d832

Here's when deodexing (using x):
https://gist.github.com/41298bc5381c199feb502b9167f514af

And here's the expected output (when using 2.0.5):
https://gist.github.com/2a09ea663042a6497d8a0c8326ec193e

The full command I am using is:
java -jar -Xmx1024M baksmali-2.2b4.jar x -o core -d . core.odex

It seems to think that p1 is an instance of Ljava/lang/reflect/Field; even though it's passed as a Ljava/lang/Object; and there doesn't seem to be any cast in the method.

**Comment by JesusFreke** [comment]
Oh, sorry. I see what you were doing there :). I'll take a deeper look this evening/this weekend.

**Comment by JesusFreke** [comment]
Also, just to confirm - this is on a device with dalvik (not art), correct?

I had made some deodexing changes that were required for art, and I thought would be safe on dalvik, but apparently not.

**Comment by HomerSp** [comment]
Yeah, it's dalvik :)

**Comment by JesusFreke** [comment]
I believe this should be fixed by adb12356c30ee61b4585530b7c31e1e7e0eff349

**Comment by HomerSp** [comment]
It is indeed, thank you for the quick fix! :)
